sygnal/helper/proxy/proxy_asyncio.py

```diff
@@ -0,0 +1,241 @@
+from asyncio import BaseTransport
+from asyncio.futures import Future
+from asyncio.transports import Transport
+from base64 import urlsafe_b64encode
+from ssl import SSLContext
+from typing import Callable, Optional, Union
+
+from sygnal.exceptions import ProxyConnectError
+from sygnal.helper.proxy import decompose_http_proxy_url
+
+logger = logging.getLogger(__name__)
+
+
+class HttpConnectProtocol(asyncio.Protocol):
+    """
+    This is for use with asyncio's Protocol and Transport API.
+
+    It performs the setup of a HTTP CONNECT proxy connection, then the calling
+     code is responsible for handing over to another asyncio.Protocol.
+
+    For Twisted, see twisted_connectproxyclient.py instead.
+
+    The intended usage of this class is to use it in a protocol factory in
+    `AbstractEventLoop.create_connection`, then await `wait_for_establishment`
+    and hand the transport over to another protocol, potentially wrapping it in
+    TLS with `AbstractEventLoop.start_tls`.
+
+    Once the connection is made, the `HttpConnectProtocol` is redundant and can
+    be forgotten about; the protocol stack might look like:
+
+                before                           after
+                                                 +------------------+
+                                                 |  HTTP Protocol   |
+                +---------------------+          +------------------+
+                | HTTP Proxy Protocol |   ===>   | SSL/TLS Protocol |
+                +---------------------+----------+------------------+
+                |              Underlying TCP Transport             |
+                +---------------------------------------------------+
+        (assuming a proxied HTTPS connection is what you were after)
+    """
+
+    def __init__(
+        self, proxy_url_parts, target_hostport: str,
+    ):
+        """
+        Args:
+            proxy_url_parts (ParseResult):
+                The URL of the HTTP proxy after being parsed by urlparse.
+                Used in the `Host` request header to the proxy, and for the
+                extraction of basic authentication credentials (if required).
+
+            target_hostport (str):
+                The host & port of the destination that the proxy should connect
+                to on your behalf. Must include a port number.
+                Examples: 'example.org:443'
+        """
+        self.completed = False
+        self.target_hostport = target_hostport
+        self.buffer = b""
+        self.transport: Transport = None  # type: ignore
+        self.proxy_url_parts = proxy_url_parts
+
+        # This future is completed when it is safe to take back control of the
+        # transport (which is also returned to indicate this).
+        self.wait_for_establishment: Future[Transport] = Future()
+
+    def data_received(self, data: bytes) -> None:
+        super().data_received(data)
+        self.buffer += data
+        if b"\r\n\r\n" in self.buffer:
```

Comment:
as a general rule: I'm a fan of early return in this sort of case. It leads to less indentation, and, more importantly, less mental load (I don't need to scroll all the way down to find out what happens in the `else` case; when I'm reading the rest of this function, I know that we're just handling the case that we've read all the headers).

Alternatively, factor out the whole long `if` body into a separate function, but early return is generally easier.

```suggestion
        if b"\r\n\r\n" not in self.buffer:
            # haven't finished the headers yet
            return
```
```

Reply:
Very true, wish I spotted these things easier :)